### PR TITLE
ACCUMULO-4461: modified ClientOpts to prompt for missing password

### PR DESCRIPTION
**Pull request by milleruntime**
Created promptUser() to prompt the user on the console for a password if one is missing. If System.console is null, it will return null causing getToken() to return null (like it does currently).

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java [hunk 1]
```diff
@@ -92,6 +94,20 @@ public Password(String dfault) {
...
     }
+
+    /**
+     * Prompts user for a password
+     *
+     * @return user entered Password object, null if no console exists
+     */
+    public static Password promptUser() throws IOException {
+      if (System.console() == null) {
+        return null;
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java]: I'm thinking that throwing an exception here would be more clear. This method cannot properly function as the caller expected, so they should get a clear exception in this case (not just a null return value).

File: core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java [hunk 2]
```diff
@@ -142,11 +158,22 @@ public AuthenticationToken getToken() {
...
+    } else if (password != null) {
+      pass = password;
+    } else {
+      try {
+        pass = Password.promptUser();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+    if (pass != null) {
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java]: This becomes simpler too. Either we throw an RTE from the attempt to get a password, or we create a `PasswordToken` with the results. This method always gets to return a non-null value (woo!).

Comment (milleruntime) [inline core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java]: I had something like that before running tests. [TestClientOpts.test ](https://github.com/apache/accumulo/blob/master/core/src/test/java/org/apache/accumulo/core/cli/TestClientOpts.java#L69) expects for that method and ones that call it to return null.

Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java]: That's a weird test. I feel like the `fail()` is the expected execution path (e.g. it's asserting that `getPrincipal()` throw an exception) and the `assertNull()` is an afterthought?

Comment (milleruntime) [inline core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java]: OK I like removing the null return. If I get rid of the try/catch, I'd have to make getToken() throw IOException which will touch a bunch of files.

Comment (milleruntime) [inline core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java]: I will keep the try catch but remove the check for null

File: core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java (hunk 1 above)
Comment (milleruntime) [inline core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java]: Fixed in 895591c

**Comment by joshelser** [comment]
This looks good to me. I will try to find some time to test and merge. Thanks Mike!

**Comment by joshelser** [comment]
> I will try to find some time to test and merge

Started running a `verify -Psunny`. Will get this in today.

**Comment by joshelser** [comment]
Uh oh. Looks like we have a failure in ReadWriteIT#sunnyDay()

```
Tests run: 8, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 160.031 sec <<< FAILURE! - in org.apache.accumulo.test.functional.ReadWriteIT
sunnyDay(org.apache.accumulo.test.functional.ReadWriteIT)  Time elapsed: 14.882 sec  <<< ERROR!
java.io.IOException: Failed to run `accumulo admin stopAll`
	at org.apache.accumulo.test.functional.ReadWriteIT.sunnyDay(ReadWriteIT.java:169)
```

```
2016-09-21 12:18:57,587 [util.Admin] ERROR: java.io.IOException: Attempted to prompt user on the console when System.console = null
java.lang.RuntimeException: java.io.IOException: Attempted to prompt user on the console when System.console = null
        at org.apache.accumulo.core.cli.ClientOpts.getToken(ClientOpts.java:171)
        at org.apache.accumulo.server.util.Admin.execute(Admin.java:213)
        at org.apache.accumulo.server.util.Admin.main(Admin.java:154)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.apache.accumulo.start.Main$2.run(Main.java:157)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: Attempted to prompt user on the console when System.console = null
        at org.apache.accumulo.core.cli.ClientOpts$Password.promptUser(ClientOpts.java:105)
        at org.apache.accumulo.core.cli.ClientOpts.getToken(ClientOpts.java:169)
        ... 8 more
```

Kind of seems like we don't have a good understanding of commands/tools which need a username + credentials (password) and which don't. Maybe that's how we need to address the problem?

**Comment by milleruntime** [comment]
Similar to the test error, this change also forces the user to enter password even when running stop-all.sh.  Which is not necessary and cumbersome.

**Comment by milleruntime** [comment]
@joshelser check out 52e5444e1657d859a0c76d89c823f100e03d57ff I tested the Admin commands and I couldn't find any that required password or auth token. This looks to be the only call to getToken that checks for null.

**Comment by joshelser** [comment]
Looks reasonable. Let's see what https://builds.apache.org/job/Accumulo-Pull-Requests/449/console does. It looks like it's doing the right thing. I tweaked this earlier this week to do a normal `mvn verify -Psunny`. Hopefully, I don't have to re-validate things by hand :)

**Comment by milleruntime** [comment]
Ok cool. I ran mvn verify -Psunny locally but coudln't figure out why tests were failing on ShellServerIT trace https://github.com/apache/accumulo/blob/1.7/test/src/test/java/org/apache/accumulo/test/ShellServerIT.java#L1444

**Comment by joshelser** [comment]
> coudln't figure out why tests were failing on ShellServerIT trace

Yeah, that one is flaky. Don't worry about it :)

**Pull request closed without merging**